prov/sm2/src/sm2_ep.c

```diff
@@ -0,0 +1,1024 @@
+	.close = sm2_srx_close,
+	.bind = sm2_srx_bind,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
+};
+
+static struct fi_ops_ep sm2_srx_ops = {
+	.size = sizeof(struct fi_ops_ep),
+	.cancel = sm2_ep_cancel,
+	.getopt = fi_no_getopt,
+	.setopt = fi_no_setopt,
+	.tx_ctx = fi_no_tx_ctx,
+	.rx_ctx = fi_no_rx_ctx,
+	.rx_size_left = fi_no_rx_size_left,
+	.tx_size_left = fi_no_tx_size_left,
+};
+
+static int sm2_ep_srx_context(struct sm2_domain *domain, size_t rx_size,
+			      struct fid_ep **rx_ep)
+{
+	struct sm2_srx_ctx *srx;
+	int ret = FI_SUCCESS;
+
+	srx = calloc(1, sizeof(*srx));
+	if (!srx)
+		return -FI_ENOMEM;
+
+	ret = ofi_spin_init(&srx->lock);
+	if (ret)
+		goto err;
+
+	sm2_init_queue(&srx->recv_queue, sm2_match_msg);
+	sm2_init_queue(&srx->trecv_queue, sm2_match_tagged);
+	sm2_init_queue(&srx->unexp_msg_queue, sm2_match_msg);
+	sm2_init_queue(&srx->unexp_tagged_queue, sm2_match_tagged);
+
+	srx->recv_fs = sm2_recv_fs_create(rx_size, NULL, NULL);
+
+	srx->min_multi_recv_size = SM2_INJECT_SIZE;
+	srx->dir_recv = domain->util_domain.info_domain_caps & FI_DIRECTED_RECV;
+
+	srx->peer_srx.owner_ops = &sm2_srx_owner_ops;
+	srx->peer_srx.peer_ops = &sm2_srx_peer_ops;
+
+	srx->peer_srx.ep_fid.fid.fclass = FI_CLASS_SRX_CTX;
+	srx->peer_srx.ep_fid.fid.context = srx;
+	srx->peer_srx.ep_fid.fid.ops = &sm2_srx_fid_ops;
+	srx->peer_srx.ep_fid.ops = &sm2_srx_ops;
+
+	srx->peer_srx.ep_fid.msg = &sm2_srx_msg_ops;
+	srx->peer_srx.ep_fid.tagged = &sm2_srx_tag_ops;
+	*rx_ep = &srx->peer_srx.ep_fid;
+
+	return FI_SUCCESS;
+
+err:
+	free(srx);
+	return ret;
+}
+
+int sm2_srx_context(struct fid_domain *domain, struct fi_rx_attr *attr,
+		    struct fid_ep **rx_ep, void *context)
+{
+	struct sm2_domain *sm2_domain;
+
+	sm2_domain =
+		container_of(domain, struct sm2_domain, util_domain.domain_fid);
+
+	if (attr->op_flags & FI_PEER) {
+		sm2_domain->srx =
+			((struct fi_peer_srx_context *) (context))->srx;
+		sm2_domain->srx->peer_ops = &sm2_srx_peer_ops;
+		return FI_SUCCESS;
+	}
+	return sm2_ep_srx_context(sm2_domain, attr->size, rx_ep);
+}
+
+static int sm2_ep_ctrl(struct fid *fid, int command, void *arg)
+{
+	struct sm2_attr attr;
+	struct sm2_domain *domain;
+	struct sm2_ep *ep;
+	struct sm2_av *av;
+	int ret, self_gid;
+
+	ep = container_of(fid, struct sm2_ep, util_ep.ep_fid.fid);
+	av = container_of(ep->util_ep.av, struct sm2_av, util_av);
+
+	switch (command) {
+	case FI_ENABLE:
+		if ((ofi_needs_rx(ep->util_ep.caps) && !ep->util_ep.rx_cq) ||
+		    (ofi_needs_tx(ep->util_ep.caps) && !ep->util_ep.tx_cq))
+			return -FI_ENOCQ;
+		if (!ep->util_ep.av)
+			return -FI_ENOAV;
+
+		attr.name = ep->name;
+		attr.flags = ep->util_ep.caps & 0;
+
+		ret = sm2_create(&sm2_prov, &attr, &av->mmap, &self_gid);
+		ep->gid = self_gid;
+
+		if (ret)
+			return ret;
+
+		if (!ep->srx) {
+			domain = container_of(ep->util_ep.domain,
+					      struct sm2_domain,
+					      util_domain.domain_fid);
+			// TODO Figure out what size goes here?? ep->tx_size?
```

Comment:
srx is "shared receive context" so it should be the ep->rx_size

Reply:
done